### PR TITLE
ENH/API Construction with decorator & attribute propagation

### PR DESCRIPTION
**Pull request by caspervdw**
This continues the work of @danielballan in https://github.com/soft-matter/slicerator/pull/4, it adds a decorator to easily implement `Slicerator` in classes.

Also it fixes some issues with attribute propagation. It doesn't work with `__getattr__` anymore; propagated attributes are actually set to the Slicerator with `setattr` at construction. Attributes that are indexed (`slice_attrs`) will be wrapped applying `_map_index` to the index.

```
class Dummy(object):
    def __init__(self):
        self.frame = list('abcdefghij')
        self._time = list(range(len(self.frame)))
        self._filename = 'filename'

    def __len__(self):
        return len(self.frame)

    @slicerate(slice_attrs=['time'], expose_attrs=['filename'])
    def __getitem__(self, i):
        return self.frame[i]  # actual code of get_frame

    def time(self, i):
        return self._time[i]

    @property
    def filename(self):
        return self._filename
```

### CODE REVIEW

**Comment by danielballan** [comment]
I think I like the decorator-style construction. That does seem simpler, and it still allows for the possibility of making a Slicerator out of a simple function.

I definitely like `slice_attrs`, making sliceable attributes explicit instead of automatic (and, therefore, at risk of being accidental or wrong).

**Comment by caspervdw** [comment]
It also fixes the pipeline issue of https://github.com/soft-matter/pims/pull/183, btw

There is one consideration still. would you prefer this over the last example?

In pims for example, this could get subclassed readers (for instance the Cine reader) to easily show which fields to propagate, without redefining `__getitem__` and the full list of exposed attrs.

```
class Dummy(object):
def __init__(self):
    self.frame = list('abcdefghij')
    self._time = list(range(len(self.frame)))
    self._filename = 'filename'

def __len__(self):
    return len(self.frame)

@slicerate
def __getitem__(self, i):
    return self.frame[i]  # actual code of get_frame

@propagate_indexed
def time(self, i):
    return self._time[i]

@propagate
@property
def filename(self):
    return self._filename
```

File: slicerator.py [hunk 1]
```diff
@@ -108,7 +172,8 @@ def __getitem__(self, key):
             rel_indices = range(start, stop, step)
             new_length = len(rel_indices)
             indices = _index_generator(rel_indices, abs_indices)
-            return Slicerator(self._ancestor, indices, new_length)
+            return Slicerator(self._ancestor, '__getitem__', indices, new_length,
```
Comment (caspervdw) [inline slicerator.py]: @danielballan This keeps bugging me: why is the first argument `self._ancestor` and not `self`?

File: slicerator.py [hunk 2]
```diff
@@ -138,25 +205,20 @@ def __getitem__(self, key):
...
+        # Any exposed attrs are lost.
         return [self._get(key) for key in self.indices]
 
     def __setstate__(self, data_as_list):
         # When deserializing, restore the Slicerator
-        return self.__init__(data_as_list)
+        return self.__init__(data_as_list, '__getitem__')
 
     def close(self):
         "Closing this child slice of the original reader does nothing."
```
Comment (caspervdw) [inline slicerator.py]: obsolete

**Comment by danielballan** [comment]
Hmm. It's a question of whether the parent or the child (the subclass-ed or the subclass-er) gets to decide what propages.

**Comment by nkeim** [comment]
Big fan of `@propagate_indexed`. `@propagate` is a difficult choice because it means that only methods or properties can be propagated, and that it must be explicit. What are the advantages of that approach over pims' own current approach of just propagating all attributes as non-slicerated attributes? That carried the advantage of an implicit model, that child objects were nearly work-alikes of the original.

With either approach: how do we deal with methods like `close()`, where it's desirable for parents and children to have distinctly different behavior? I'm recalling the saga of https://github.com/soft-matter/pims/pull/147 .

**Comment by caspervdw** [comment]
@danielballan I think that the subclassed reader should decide on which fields to propagate, as it generally contains the attributes (for pims at least..). But the subclasser can also decide to propagate a field if we use the `@propagate` approach, they both can decide.

Generally about propagating fields (@nkeim 's comment): I think it is a matter of how you see the sliced object. If you look at `frames[:10].close()`, what would you expect to happen? If we pursue a thin wrapper that 'just makes slices sliceable', I would not find it strange that  `frames[:10].close()` closes the whole `frames` object, taking al the slices with it.

With other words, if we want to make just a transparent layer adding a functionality to slicing, maybe we should try to let Slicerator mimic the ancestor as close as possible, forwarding every attribute (and adding as less attributes as possible), although it may lead to some unexpected issues (in which unexpected of course is just a point of view)

So in light of this I tend to agree with Nathan, propagating every field and just be clear about that.

Finally, I think that in the way it is implemented now, the `method` parameter could default to `__getitem__` or it may even be dropped completely because Slicerator is supposed to mimic the ancestor object which makes `__getitem__` the only logical choice. (the use with another method would be `frames.other_method(slice(10))[0]`)

**Comment by danielballan** [comment]
Blacklisting attributes is convenient, as long as users understand what is happening. Whitelisting attributes is safer. I am strongly against _always_ propagating all fields: at minimum we should allow the user to blacklist, perhaps with `@do_not_propagate` (with a better name). I can see the case for going with a blacklist instead of our current whitelist approach.

**Comment by nkeim** [comment]
1. I like the case for whitelisting as the default. More Pythonic. We can always add a blacklist option later, but for this PR, YAGNI.
2. @caspervdw has a good point for the `close()` issue — it's really not the end of the world as far as pims is concerned. But I worry that Slicerator users outside pims might not see it that way.

A possible way to address both (1) and (2) architecturally is to subclass Slicerator. `@propagate_indexed` etc. can be classmethods. Then slicerator can include `SliceratorWorkAlike`, and pims can define `SliceratorWithClose`.

**Comment by caspervdw** [comment]
Agreed. Let's supply the whitelist option and we might make other versions later.

This is actually as it is implemented. I extended it with the decorators `@propagate` and `@propagate_indexed`. I can't get the decorator `@propagate` to work with properties.

Anyone in favor of putting `'__getitem__'` as default method?

And finally, if we would transfer the whole key -> indices logic from `__getitem__`, to `__init__`, then this would enable `@slicerate` to not make 2 Slicerators but only one.

**Comment by nkeim** [comment]
`__getitem__` should definitely be the default method when constructing from an existing object. Is that what you mean?

In your second API example, how will the whitelist be specified? Maybe with a class attribute?

**Comment by caspervdw** [comment]
@nkeim I combined the two APIs, see the last test in test.py. Fields will be propagated if they are listed in the whitelist _or_ if they are decorated with `@propagate` or `@propagate_indexed`.

Indeed, I meant making `__getitem__` the default attribute for the `method` parameter.

My last point was moving the key - indices logic. But I think I will leave that for later, also in view of the refactoring that might be going to happen due to #6

**Comment by nkeim** [comment]
I really like this API!

If `__getitem__` becomes the default constructor parameter, I'll just advocate one more time for dropping `from_list()`. Or maybe rename it to `from_sequence()`? It's just weird to name such a specific Python type in the function, when the beauty of slicerator is that it can wrap any object with indices and a length, and that's its primary mode of operation. @caspervdw, it sounds like you may have to break the tie on this one.

If we _really_ wanted a clean API, we could move all of the internal-use options from `Slicerator.__init__()` to `Slicerator._spawn()`.

**Comment by caspervdw** [comment]
I agree with Nathan to drop `from_list` if we make __getitem_ the default method. Wrapping a list would be nothing more than `Slicerator(list)` if I am not mistaken.

I leave the refactoring for later, I think there is indeed room for an even cleaner class.

Anyone wants to give it a try to make @propagate work with properties?

**Comment by danielballan** [comment]
Chiming in quickly to say, don't take my silence as unhappiness with the direction this is going. Just too swamped at work to think about it for the next day or two.

**Comment by nkeim** [comment]
No problem! 

This is what happens when an entire package is basically one class :).

**Comment by caspervdw** [comment]
I fixed the last things, additonaly made propagation of attributes possible with pipelines.

**Comment by caspervdw** [comment]
Ok, this approach has its shortcomings. There is really no way to have a propagated attribute reference `self` to the actual Slicerator. The method stays bound to the ancestor so `_indices` or `__len__` are not accessible.

I can't get the following method (from the norpix_reader) to propagate properly.

```
def dump_times_float(self):
    return np.array([self._get_time(i)[0] for i in range(len(self))])
```

Because `self` stays the ancestor.

**Comment by nkeim** [comment]
Yeah. This has come up before, and was the motivation for [my context-based scheme](https://github.com/soft-matter/pims/pull/160) that's currently in pims. [In a comment](https://github.com/soft-matter/pims/issues/159#issuecomment-120762879), @danielballan suggested the way to avoid such an elaborate, "clever" solution: more Slicerators! It seemed like a good idea to me.

**Comment by caspervdw** [comment]
Right, so then we need yet another category of propagated metadata?
- 'normally' propagate (`propagate`)
- propagate & reindex (`propagate_indexed`)
- propagate & slice (`propagate_sliced`?)

I am starting to come round and think a metaclass might be the best method after all. I never used that but it seems to do what we would need: modify an existing class to become a slicerator. Subclassing might be possible too, but might be tricky in combination with classes that are already subclasses.

Then a to-be slicerator class could define fields `propagate`, `propagate_indexed`, `propagate_sliced` which the Slicerator could sort out.

What do you think is the best way to go?

**Comment by caspervdw** [comment]
We might also change the API to stick with `propagate` and `propagate_indexed` only. This creates an alternative for `dump_times_float`. This could then be reached with `get_time[:]`.

```
sliced = reader[1:]  # returns a Slicerator
sliced.attribute # returns reader.attribute
sliced.get_time(i, **kwargs)  # reindexes and calls reader.get_time(map_index(i)), **kwargs)
sliced.get_time(slice(1, 10), **kwargs)  # gives you iterable
sliced.get_time[i]  # calls reader.get_time(map_index(i))
sliced.get_time[1:10]  # gives an iterable iterating over reader.get_time
```

This can be done by wrapping the `get_time` inside a class having both `__call__` and `__getitem__` methods. You only need the fancy indexing part inside the `__getitem__`; it doesn't need to return a Slicerator.

**Comment by caspervdw** [comment]
I implemented the ideas in the previous post; please ignore the rest of my random-walk towards this result. Adding to @danielballan s overview in the previous PR:

## Construction of Slicerator objects
1. From any object that supports `__getitem__` and `__len__`
   
   ```
   Slicerator([1,2,3])
   ```
2. From a function (any callable) that accepts an integer argument.
   The length must be specified.
   
   ```
   Slicerator.from_func(func, 10)
   ```
3. From a class that defines `__getitem__` and `__len__`:
   
   ```
   cls = Slicerator.from_class(some_class)
   my_object = cls()
   ```
4. In a class definition (which is essentially the same as 3):
   
   ```
   @Slicerator.from_class
   class Dummy(object):
       def __init__(self):
           self.frame = list('abcdefghij')
   
       def __len__(self):
           return len(self.frame)
   
       def __getitem__(self, i):
           return self.frame[i]
   ```

To control whether attributes are exposed through subclasses of a slicerator class, there are different ways to whitelist attributes. 
1. Any of the constructors take an optional argument, `propagated_attrs`, which is a whitelist of attributes that will be propagated through the Slicerator and its children. This overwrites any other propagation list.
2. Methods can be decorated with `@Slicerator.propagate_attr`, in which the method gets a propagation 'hook' that will be present in derived subclasses (unless the method is overwritten)
3. A class can define a list of attributes that will be propagated using the `propagate_attrs` field. This field can be completely redefined by subclasses.

And then there is the `@Slicerator.indexed_attr` for methods that take an index as their first argument. It gives these methods fancy indexing capabilities, and also makes sure that the index is mapped correctly if the method is propagated. This is done through a new class `SliceableAttribute`, which is a rip-off of `Slicerator`.

The example below illustrates the current metadata propagation & class constructor (example is tested in `test_getattr_subclass`)

What do you think of the current API? @danielballan @tacaswell @nkeim 
I would prefer to close this PR soon and defer other enhancements to other PRs.

```
@Slicerator.from_class
class Dummy(object):
    propagated_attrs = ['attr1']
    def __init__(self):
        self.frame = list('abcdefghij')

    def __len__(self):
        return len(self.frame)

    def __getitem__(self, i):
        return self.frame[i]

    def attr1(self):
        # propagates through slices of Dummy
        return 'sliced'

    @Slicerator.propagate_attr
    def attr2(self):
        # propagates through slices of Dummy and subclasses
        return 'also in subclasses'

    def attr3(self):
        # does not propagate
        return 'only unsliced'


class SubClass(Dummy):
    propagated_attrs = ['attr4']  # overwrites propagated attrs from Dummy

    def attr4(self):
         # propagates through slices of SubClass
        return 'only subclass'
```

**Comment by nkeim** [comment]
Two thumbs up! :+1: :+1:  My only quibble is that `propagated_attrs` in the class definition should begin with an underscore.

File: slicerator.py [hunk 3]
```diff
@@ -88,80 +226,85 @@ def __iter__(self):
...
+            if hasattr(attr, '_indexed'):
+                return SliceableAttribute(self, getattr(self._ancestor, name))
+            else:
+                return getattr(self._ancestor, name)
+        raise AttributeError
 
     def __getstate__(self):
         # When serializing, return a list of the sliced and processed data
+        # Any exposed attrs are lost.
         return [self._get(key) for key in self.indices]
```
Comment (nkeim) [inline slicerator.py:260]: I'm not sure whether this is the right default behavior. A major use case for Slicerators is for sequences that are too large to fit in memory or very expensive to compute. So serialization should probably get its own issue.

File: slicerator.py [hunk 4]
```diff
@@ -48,24 +98,101 @@ def __init__(self, ancestor, indices=None, length=None):
...
+            def __getitem__(self, i):
+                return func(i)
+
+            def __len__(self):
+                return length
+
+        return cls(Dummy(), propagated_attrs=propagated_attrs)
+
+    @classmethod
+    def from_class(cls, other_class, propagated_attrs=None):
```
Comment (nkeim) [inline slicerator.py]: This and a few other methods need docstrings. I'd be OK with leaving them out of this PR, though.

File: slicerator.py (hunk 3 above)
Comment (caspervdw) [inline slicerator.py:260]: Right, I basically implemented this to get IPython mulciprocessing to work (which, by the way, isn't tested yet, see issue on trackpy). I thought that a file reference in general is not picklable. A seperate issue could definitely improve on this.

File: slicerator.py (hunk 4 above)
Comment (caspervdw) [inline slicerator.py]: I'll write them if everybody agrees on this API, probably based on my previous post on this issue. They can go in this PR.

About this specific constructor: would you rather subclass `other_class` into a new class and return that? Currently, `other_class` will be changed in-place in.

**Comment by danielballan** [comment]
Tom and I are sitting down at work today to take a hard look at this. Apologies again for the delay. The X-ray beam turned back on this week....

File: slicerator.py (hunk 4 above)
Comment (danielballan) [inline slicerator.py]: I think a subclass would be safer. Can anyone think of an advantage in _not_ subclassing?

Comment (nkeim) [inline slicerator.py]: Agreed, and I can't think of an advantage. `isinstance` will still work, which is what matters.

**Comment by danielballan** [comment]
OK, I finally caught up. I really like where this landed. I think we can run with this API. I agree we should merge this general API (with docstrings) and resolve other issues separately.

Some quibbles:
- I'm not used to seeing utility decorators like `propage_attr` and `indexed_attr` as class methods. Maybe they should be top-level functions in the `slicerator` module instead.
- Can you summarize why we need `SliceableAttribute`? Could these not just be more `Slicerator` objects?

**Comment by danielballan** [comment]
One more: I totally concede that `from_list` is not necessary now. It feels a little weird to expose a private method, `__getitem__` as a default argument, but that's hard to avoid given what we are doing.

**Comment by caspervdw** [comment]
@danielballan @nkeim I addressed your comments:
- `from_class` now subclasses and makes sure that the resulting class looks like the original. It does add the hook `_slicerator_hook` to identify the slicerator.
- renamed `propagated_attrs` to `propagate_attrs`
- renamed Slicerator attribute `propagate_attrs` to `_propagate_attrs`
- decorators `index_attr` and `propagate_attr` are now top-level methods
- added docstrings

The `SliceableAttribute` functions different from the `Slicerator`. It supports `__call__`, so that you can do `sliced.get_time(i, **kwargs)` and it does not spawn any other Slicerators (sliced.get_time[:] gives you a generator).

Also, I _dropped_ the whole method attribute (instead of making it  `__getitem__` by default), because it leaded to some issues (which I sadly cannot remember). I figured that if we really wanted that, it could be a separate PR. But I like to see the `Slicerator` as something that 'just' extends the use of `__getitem__`.

**Comment by danielballan** [comment]
Looks good to me. I like the added tests.

I see some minor PEP8 issues and docstring edits, but I'll save them for a future round.

I leave it to @nkeim to take another look and merge if he's happy.

File: slicerator.py [hunk 5]
```diff
@@ -48,23 +60,134 @@ def __init__(self, ancestor, indices=None, length=None):
...
+        else:
+            # check propagated_attrs field from the ancestor definition
+            if hasattr(ancestor, 'propagate_attrs'):
+                self._propagate_attrs = ancestor.propagate_attrs
+            else:
+                self._propagate_attrs = []
+
+            # add methods having the _propagate hook
+            for name in dir(ancestor):
+                if hasattr(getattr(ancestor, name), '_propagate_hook'):
```
Comment (caspervdw) [inline slicerator.py]: this actually evaluates _all_ properties of `ancestor`. Not desired behaviour, I think. Need a workaround here.

**Comment by caspervdw** [comment]
@danielballan @nkeim sorry that I keep pushing changes; I ran into some issues while rebasing pims. Also I solved #6 pipeline propagation & nesting.

**Comment by danielballan** [comment]
@nkeim Do you have a moment this weekend to take a last look at this?

**Comment by nkeim** [comment]
It's great! I'm sure we'll want to make some tweaks as we build things on it. But I think it's time to get started with that.

I'll leave you with a semantic question: I always thought a "hook" was an extension point in code, where you could insert a reference to your own custom function. (This idea is often used in formatting or pretty-printing code.) Is that true? The use here seems more like a "flag." But since it's all private here, it really doesn't matter!

Great job, @caspervdw and all.

**Comment by caspervdw** [comment]
@danielballan Do you have time to tag a release so that pims tests can pass?

**Comment by tacaswell** [comment]
:+1: good job!

**Comment by danielballan** [comment]
Yes, I'll tag a release so we can get a pims release out ASAP.

I'm sure there are little things to change, but it will easier to work from a stable starting point.

I had the same reaction as @nkeim about "hook" vs. "flag"... we can open a separate PR.

**Comment by danielballan** [comment]
Slicerator v0.9.3 is on PyPI.
